tmt/steps/__init__.py

```diff
@@ -267,32 +264,45 @@ def setup_actions(self) -> None:
                 f"with order '{reboot_plugin.order}'.", level=2)
             self._phases.append(reboot_plugin)
 
+    @overload
+    def phases(self, classes: None = None) -> List[Phase]:
+        pass
+
     @overload
     def phases(self, classes: Type[PhaseT]) -> List[PhaseT]:
         pass
 
     @overload
-    def phases(self, classes: None = None) -> List[Phase]:
+    def phases(self, classes: Tuple[Type[PhaseT], ...]) -> List[PhaseT]:
```

Comment:
This line seems to be wrong. `Tuple[Type[PhaseT], ...]` means a tuple with the exactly same elements in it (derived from `Phase`), e.g. `(Plugin, Plugin, Plugin)`...

The intended behavior on this method should be to allow a tuple with a set of different types which would return a list of values of types from the same set. E.g. calling `phases(classes=(ReportPlugin, FinishPlugin))` would return `List[Union[ReportPlugin, FinishPlugin]]`. Although in the code, it is never used with a tuple, the `classes` parameter is always `None` or a single class.

Reply:
> This line seems to be wrong. `Tuple[Type[PhaseT], ...]` means a tuple with the exactly same elements in it (derived from `Phase`), e.g. `(Plugin, Plugin, Plugin)`...
> 
> The intended behavior on this method should be to allow a tuple with a set of different types which would return a list of values of types from the same set. E.g. calling `phases(classes=(ReportPlugin, FinishPlugin))` would return `List[Union[ReportPlugin, FinishPlugin]]`.

I see. Yes, the intended behavior is the one of return value matching given `classes`. And AFAICT, the specifications seem to be correct but linters are slightly behind: the language server in Code decodes this correctly, showing return value of `self.phases(classes=(Action, ReportPlugin))` to be `Action | ReportPlugin`, but mypy reports `Phase`, which is wrong.

There's not much we can do about that, I'm afraid, I can add a big fat TODO comment and try to refactor it later.

> Although in the code, it is never used with a tuple, the `classes` parameter is always `None` or a single class.

I disagree: a single class https://github.com/teemtee/tmt/pull/1365/files#diff-d9422da1a3efb9209a7d4beb7d824f85a19e5abe91fb239f74806847d88db266R304, a tuple of classes e.g. https://github.com/teemtee/tmt/pull/1365/files#diff-5f59acdec9a1fdc81c556e6ca1c7c872a25d72f1be0259a8f727b9dce4a45a44R140, https://github.com/teemtee/tmt/pull/1365/files#diff-458194650925d6c3ea27ee8a599db67a78a23dbb0b353cddc9e2b8387b54d993R159, https://github.com/teemtee/tmt/pull/1365/files#diff-bc07f2874fc5fd6a1ce91cb1d3d2ffd2dca74667815cc25dfff00e40dd0d17f5R75 - basically every step is calling `phases(classes=(Action, StepSpecificPluginBaseClass))`

Reply:
Alright, thanks for the explanation. Not sure what would be the right solution at the moment.

> I disagree: a single class ... - basically every step is calling phases(classes=(Action, StepSpecificPluginBaseClass))

I see, I missed that somehow. Thanks for the clarification.